###Author: Harry Fischl
###Converts hg19 aligned bam file to bigwig files of reads narrowed to their 3' nucleotide
##########Processing BAM FILE to Narrowed BIGWIG Files

TWD1 = "/path/to/bam/file/directory/"
TWD2 = "/path/to/bigwig/file/directory/"
setwd(TWD1)

library("BSgenome.Hsapiens.UCSC.hg19")
seq_info_object1 = SeqinfoForBSGenome(Hsapiens)
library(GenomicAlignments)
library(rtracklayer)
###bam files generated by alignment of fastq files to the hg19 genome using the Ion Torrent Server TMAP aligner with default alignment settings (-tmap mapall stage1 map4)

blf1 = list.files(pattern=".bam")

for(a1 in 1:length(blf1)){
  ##############################LOAD BAM FILE
  bamFile = blf1[a1]
  print(bamFile)
  bamFileName = strsplit(bamFile, ".bam")[[1]]
  greads = readGappedReads(bamFile)
  gr1 = as.data.frame(greads)
  ##############################NARROW EACH READ TO 3' NUCLEOTIDE
  gr1$pA <- ifelse(gr1$strand == "-",gr1[,"start"],gr1[,"end"])
  gr1$start = gr1$pA
  gr1$end = gr1$pA
  keeps <- c("seqnames","strand","start","end")
  gr2 = gr1[,keeps]
  ###########################BIGWIG FILES
  gr2pos <- gr2[gr2$strand == "+",]
  gr2neg <- gr2[gr2$strand == "-",]
  ##convert GRanges
  gr3pos <- makeGRangesFromDataFrame(gr2pos, keep.extra.columns=TRUE, ignore.strand=FALSE, seqinfo=seq_info_object1, seqnames.field="seqnames", start.field="start", end.field="end", strand.field="strand", starts.in.df.are.0based=FALSE)
  gr3neg <- makeGRangesFromDataFrame(gr2neg, keep.extra.columns=TRUE, ignore.strand=FALSE, seqinfo=seq_info_object1, seqnames.field="seqnames", start.field="start", end.field="end", strand.field="strand", starts.in.df.are.0based=FALSE)
  ##convert to Coverage (bedgraph)
  covvecp <- coverage(gr3pos)
  covvecn <- coverage(gr3neg)
  ####################################OUTPUT AS BIGWIGS (1 for EACH STRAND)
  posbw = "000_pos.bw"
  negbw = "000_neg.bw"
  posbw1 = sub("000", bamFileName, posbw)
  negbw1 = sub("000", bamFileName, negbw)
  setwd(TWD2)
  export(covvecp, con=posbw1)
  export(covvecn, con=negbw1)
  ####SWITCH BACK TO BAM FILE DIRECTORY
  setwd(TWD1)
}


