###Author: Harry Fischl
###Converts hg19 and dm6 combined genome aligned bam file (from drosophila cell spiked in experiments) to bigwig files (2 per hg19 and dm6 genomes) of reads narrowed to their 3' nucleotide
###########################SEPARATING COMBINED BAM FILE
TWD1 = "/path/to/spikedin/bamfiles/directory"
TWD2 = "/path/to/human/bigwig/file/directory"
TWD3 = "/path/to/drosophila/bigwig/file/directory"

setwd(TWD1)

########################LIBRARIES
library("BSgenome.Hsapiens.UCSC.hg19")
library("BSgenome.Dmelanogaster.UCSC.dm6")

seq_info_object1 = SeqinfoForBSGenome(Hsapiens)
seq_info_object2 = SeqinfoForBSGenome(Dmelanogaster)
library(GenomicAlignments)
library(rtracklayer)
######################################################
######################################################
###bam files generated by alignment of fastq files to a combined hg19 and dm6 genome build using the Ion Torrent Server TMAP aligner with default alignment settings (-tmap mapall stage1 map4)

###Drosophila genome (dm6) chromosome names
###Some chromosomes were renamed prior to creation of combined genome to avoid clashes with hg19 names
dm6_names <- read.delim('dm6_names.txt', header = F)
dm6_names$V1 = as.vector(dm6_names$V1)
dm6nam = as.vector(sapply(dm6_names$V1, function(n) strsplit(n, ">")[[1]][2]))

###Human genome (hg19) chromosome names
hg19_names <- read.delim('hg19_names.txt', header = F)
hg19_names$V1 = as.vector(hg19_names$V1)
hg19nam = as.vector(sapply(hg19_names$V1, function(n) strsplit(n, ">")[[1]][2]))

######################################################

blf1 = list.files(pattern = ".bam$")

for(a1 in 1:length(blf1)){
  bamFile = blf1[a1]
  print(bamFile)
  bamFileName = strsplit(bamFile, "_tmaphg19dm6sorted.bam")[[1]]
  ##############################INPUT BAM FILE
  greads1 = readGappedReads(bamFile)
  ##NARROW TO 3' END OF READ
  gr1 = as.data.frame(greads1)
  gr1$pA <- ifelse(gr1$strand == "-",gr1[,"start"],gr1[,"end"])
  gr1$start = gr1$pA
  gr1$end = gr1$pA
  keeps <- c("seqnames","strand","start","end")
  gr2 = gr1[,keeps]
  gr2$seqnames = as.vector(gr2$seqnames)
  ##SPLIT BETWEEN GENOMES
  grhg19 = gr2[gr2$seqnames %in% hg19nam,]
  grdm6 = gr2[gr2$seqnames %in% dm6nam,] 
  ###CORRECT droso NAMES
  grdm6$seqnames[grdm6$seqnames == "chrMDS"] = "chrM"
  grdm6$seqnames[grdm6$seqnames == "chr4DS"] = "chr4"
  grdm6$seqnames = gsub("chrYDS","chrY",grdm6$seqnames)
  grdm6$seqnames = gsub("chrXDS","chrX",grdm6$seqnames)
  ####CONVERT TO FACTOR
  grdm6$seqnames = factor(grdm6$seqnames)
  grhg19$seqnames = factor(grhg19$seqnames)
  #########################################
  ########################### DROSO BIGWIG FILES
  setwd(TWD3)
  grdm6pos <- grdm6[grdm6$strand == "+",]
  grdm6neg <- grdm6[grdm6$strand == "-",]
  ##convert GRanges
  grdm6posx <- makeGRangesFromDataFrame(grdm6pos, keep.extra.columns=TRUE, ignore.strand=FALSE, seqinfo=seq_info_object2, seqnames.field="seqnames", start.field="start", end.field="end", strand.field="strand", starts.in.df.are.0based=FALSE)
  grdm6negx <- makeGRangesFromDataFrame(grdm6neg, keep.extra.columns=TRUE, ignore.strand=FALSE, seqinfo=seq_info_object2, seqnames.field="seqnames", start.field="start", end.field="end", strand.field="strand", starts.in.df.are.0based=FALSE)
  ##########
  dm6covvecp <- coverage(grdm6posx)
  dm6covvecn <- coverage(grdm6negx)
  #######################################
  dm6posbw = "000_dm6_pos.bw"
  dm6negbw = "000_dm6_neg.bw"
  dm6posbw1 <- sub("000", bamFileName, dm6posbw)
  dm6negbw1 <- sub("000", bamFileName, dm6negbw)
  ###############################################OUTPUT DROSO BED FILES
  export(dm6covvecp, con=dm6posbw1)
  export(dm6covvecn, con=dm6negbw1)
  #########################################
  ######################## HUMAN BEDGRAPH FILES
  setwd(TWD2)
  grhg19pos <- grhg19[grhg19$strand == "+",]
  grhg19neg <- grhg19[grhg19$strand == "-",]
  ##convert GRanges
  grhg19posx <- makeGRangesFromDataFrame(grhg19pos, keep.extra.columns=TRUE, ignore.strand=FALSE, seqinfo=seq_info_object1, seqnames.field="seqnames", start.field="start", end.field="end", strand.field="strand", starts.in.df.are.0based=FALSE)
  grhg19negx <- makeGRangesFromDataFrame(grhg19neg, keep.extra.columns=TRUE, ignore.strand=FALSE, seqinfo=seq_info_object1, seqnames.field="seqnames", start.field="start", end.field="end", strand.field="strand", starts.in.df.are.0based=FALSE)
  ##########
  hg19covvecp <- coverage(grhg19posx)
  hg19covvecn <- coverage(grhg19negx)
  #######################################
  hg19posbw = "000_hg19_pos.bw"
  hg19negbw = "000_hg19_neg.bw"
  hg19posbw1 <- sub("000", bamFileName, hg19posbw)
  hg19negbw1 <- sub("000", bamFileName, hg19negbw)
  ###############################################OUTPUT DROSO BED FILES
  export(hg19covvecp, con=hg19posbw1)
  export(hg19covvecn, con=hg19negbw1)
  ####
  setwd(TWD1)
}




